lib/tracer/tracer.js

```diff
@@ -58,29 +69,37 @@ Tracer.prototype.start = function (name, type, {
     events: [],
     userId,
   };
-
-  return traceInfo;
 };
 
-Tracer.prototype.event = function (traceInfo, type, data, metaData) {
-  // do not allow to proceed, if already completed or errored
-  let lastEvent = this.getLastEvent(traceInfo);
+Tracer.prototype.event = function (trace, type, data, meta) {
+  const activeEvent = getActiveEvent();
 
+  // We should not nest based on the last event,
+  // but based on the active event of the current context, due to parallel execution.
+  const lastEvent = trace.events[trace.events.length - 1];
+
+  // Do not allow proceeding, if already completed or errored
   if (
-    // trace completed but has not been processed
-    lastEvent &&
-    ['complete', 'error'].indexOf(lastEvent.type) >= 0 ||
-    // trace completed and processed.
-    traceInfo.isEventsProcessed
+    // Trace completed but has not been processed
+    [EventType.Complete, EventType.Error].includes(lastEvent?.type) ||
+    // Trace completed and processed.
+    trace.isEventsProcessed
```

Comment:
Since we don't have to worry about other agents modifying the trace object, can we simplify this check?

Reply:
This one is still useful if an async event is triggered after the trace has been processed. And the first one can still prevent some corruption of the trace as far as I see.